upx-recovery-tool.py

```diff
@@ -0,0 +1,289 @@
+                self.arch = magic_arch
+                return
+
+        raise UnsupportedFileError(f"Unsupported file type '{magic_str}'")
+
+    def is_upx(self):
+        """ Method that looks for ASM signatures to identify an UPX executable """
+
+        # Instead of looking for the pattern in all the executable bytes, look for the
+        # UPX sigs in the EP. The code is not so beauty, but it'll be more efficient.
+        ep_bytes = self.get_ep_bytes(50)
+
+        for sig in self.upx_sigs[self.arch]:
+            if re.match(sig, ep_bytes, re.DOTALL):
+                return True
+
+        return False
+
+    def detect_version(self):
+        """ Method to know the UPX version used to pack the executable """
+        # TODO: The current detection method is very naive and can be easily tricked
+
+        # Load file in memory to look for string with version
+        self.in_fd.seek(0)
+        in_buff = mmap.mmap(self.in_fd.fileno(), 0, access=mmap.ACCESS_READ)
+
+        # Detect UPX version
+        #if in_buff.find(b"$Id: UPX 4.00") != -1:
+        version_off = in_buff.find(b"$Id: UPX ")
+        if version_off == -1:
+            print("[!] UPX version could not be detected")
+            self.version = None
+        else:
+            self.version = int(in_buff[version_off + 9: version_off + 10])
+            # TODO: Use 're' to get major and minor version
+
+        in_buff.close()
+
+    def load_structs(self):
+
+        eh = self.elf.header
+
+        # l_info
+        self.l_info_off = eh.e_phoff + eh.e_phnum * eh.e_phentsize
+        self.l_info = l_info_s(self.buff[self.l_info_off: self.l_info_off + 12])
+
+        # p_info
+        self.p_info_off = self.l_info_off + 12
+        self.p_info = p_info_s(self.buff[self.p_info_off: self.p_info_off + 12])
+
+    def patch(self, patch_bytes, offset):
+        """ Method to patch bytes in the output binary """
+
+        self.buff[offset: offset + len(patch_bytes)] = patch_bytes
+
+    def get_ep_bytes(self, num_bytes):
+        """ Method to get the first 'num_bytes' of the executable's Entry Point. Used to detect UPX signatures """
+        
+        ep_bytes = None
+        ep = self.elf.header['e_entry']
+
+        # Loop segments looking where the EP is
+        for seg in self.elf.iter_segments():
+            sh = seg.header
+            if ep > sh.p_vaddr and ep < sh.p_vaddr + sh.p_memsz:
+                start_off = ep - sh.p_vaddr
+                ep_bytes = seg.data()[start_off: start_off + num_bytes]
+
+        return ep_bytes
+
+    def fix(self):
+        """ Method to fix all the (supported) modifications to UPX """
+
+        try:
+            self.fix_l_info()
+
+            if self.version != 4:
+                self.fix_p_info()
+
+        finally:
+            # Close mmap
+            self.buff.flush()
+            self.buff.close()
+
+            # Close file descriptors
+            self.in_fd.close()
+            self.out_fd.close()
```

Comment:
it doesn't seem to be the right place to do the cleanup, better move it to the function where they are assigned checking if they have already been initialized